Mathlib/Order/Sublattice.lean

```diff
@@ -0,0 +1,292 @@
+
+/-- The maximum sublattice of a lattice. -/
+instance instTop : Top (Sublattice α) where
+  top.carrier := univ
+  top.supClosed' := supClosed_univ
+  top.infClosed' := infClosed_univ
+
+/-- The empty sublattice of a lattice. -/
+instance instBot : Bot (Sublattice α) where
+  bot.carrier := ∅
+  bot.supClosed' := supClosed_empty
+  bot.infClosed' := infClosed_empty
+
+/-- The inf of two sublattices is their intersection. -/
+instance instInf : Inf (Sublattice α) where
+  inf L M := { carrier := L ∩ M
+               supClosed' := L.supClosed.inter M.supClosed
+               infClosed' := L.infClosed.inter M.infClosed }
+
+/-- The inf of sublattices is their intersection. -/
+instance instInfSet : InfSet (Sublattice α) where
+  sInf S := { carrier := ⨅ L ∈ S, L
+              supClosed' := supClosed_sInter $ forall_range_iff.2 $ fun L ↦ supClosed_sInter $
+                forall_range_iff.2 fun _ ↦ L.supClosed
+              infClosed' := infClosed_sInter $ forall_range_iff.2 $ fun L ↦ infClosed_sInter $
+                forall_range_iff.2 fun _ ↦ L.infClosed }
+
+instance instInhabited : Inhabited (Sublattice α) := ⟨⊥⟩
+
+/-- The top sublattice is isomorphic to the lattice.
+
+This is the sublattice version of `Equiv.Set.univ α`. -/
+def topEquiv : (⊤ : Sublattice α) ≃o α where
+  toEquiv := Equiv.Set.univ _
+  map_rel_iff' := Iff.rfl
+
+@[simp, norm_cast] lemma coe_top : (⊤ : Sublattice α) = (univ : Set α) := rfl
+@[simp, norm_cast] lemma coe_bot : (⊥ : Sublattice α) = (∅ : Set α) := rfl
+@[simp, norm_cast] lemma coe_inf' (L M : Sublattice α) : L ⊓ M = (L : Set α) ∩ M := rfl
+@[simp, norm_cast] lemma coe_sInf (S : Set (Sublattice α)) : sInf S = ⋂ L ∈ S, (L : Set α) := rfl
+@[simp, norm_cast] lemma coe_iInf (f : ι → Sublattice α) : ⨅ i, f i = ⋂ i, (f i : Set α) := by
+  simp [iInf]
+
+@[simp, norm_cast] lemma coe_eq_univ : L = (univ : Set α) ↔ L = ⊤ := by rw [←coe_top, coe_inj]
+@[simp, norm_cast] lemma coe_eq_empty : L = (∅ : Set α) ↔ L = ⊥ := by rw [←coe_bot, coe_inj]
+
+@[simp] lemma not_mem_bot (a : α) : a ∉ (⊥ : Sublattice α) := id
+@[simp] lemma mem_top (a : α) : a ∈ (⊤ : Sublattice α) := mem_univ _
+@[simp] lemma mem_inf : a ∈ L ⊓ M ↔ a ∈ L ∧ a ∈ M := Iff.rfl
+@[simp] lemma mem_sInf {S : Set (Sublattice α)} : a ∈ sInf S ↔ ∀ L ∈ S, a ∈ L := by
+  rw [←SetLike.mem_coe]; simp
+@[simp] lemma mem_iInf {f : ι → Sublattice α} : a ∈ ⨅ i, f i ↔ ∀ i, a ∈ f i := by
+  rw [←SetLike.mem_coe]; simp
+
+/-- Sublattices of a lattice form a complete lattice. -/
+instance instCompleteLattice : CompleteLattice (Sublattice α) :=
+  { completeLatticeOfInf (Sublattice α)
+      fun _s ↦ IsGLB.of_image SetLike.coe_subset_coe isGLB_biInf with
+    bot := ⊥
+    bot_le := fun _S _a ↦ False.elim
+    top := ⊤
+    le_top := fun _S a _ha ↦ mem_top a
+    inf := (· ⊓ ·)
+    le_inf := fun _L _M _N hM hN _a ha ↦ ⟨hM ha, hN ha⟩
+    inf_le_left := fun _L _M _a ↦ And.left
+    inf_le_right := fun _L _M _a ↦ And.right }
```

Comment:
nit: use `where` notation for les `fun`

Reply:
But `where` doesn't support `with` clauses, does it?

Reply:
It does, using `__ := foo` instead of `foo with`

Reply:
Ah uh that was unexpected syntax.